readme.md
![Alt text](https://travis-ci.org/jonpchin/gochess.svg?branch=master "Travis CI Go Play Chess Image")
[![Build status](https://ci.appveyor.com/api/projects/status/96kvdw3mr190y854?svg=true)](https://ci.appveyor.com/project/jonpchin/gochess)
[![Go Report Card](https://goreportcard.com/badge/github.com/jonpchin/gochess)](https://goreportcard.com/report/github.com/jonpchin/gochess)
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](http://www.gnu.org/licenses/gpl-3.0)
-<br><br>
![Alt text](/img/gif/goplaychessdemo_shrink.gif?raw=true "Go Play Chess Demo")
<br>
![Alt text](/img/screenshots/lobbyResize.png?raw=true "Chess Lobby")
<br>
![Alt text](/img/screenshots/gameResize.png?raw=true "Chess Room")
<br><br>
Go Play Chess - Free Online Real time chess web server using websockets - Creator: Jonathan Chin<br><br>
12/10/2015 - Start date of project.<br><br>
5/31/2016 This project is officially open source. - All commit history has been cleared as some of the git history contained sensitive data. 
<br><br>
4/30/2017 I've decided to add a MUD. Chess and MUD's don't usually go together but by having them on the same web server and domain allows me to save server costs. More details will be revealed soon.
<br><br>If you are interesting in contributing please open a new issue here:
https://github.com/jonpchin/gochess/issues
In the issue state what you want to work on and I'll add you as a contributor. 
If you are open to doing anything I can assign you one of the open tasks in the back log.
Once your changes are ready, make a pull request and I"ll review it.
<br>
<br><b>Features:</b><br>
<p>1. User accounts</p>
<p>2. Time controls</p>
<p>3. Player ratings and matchmaking system </p>
<p>4. Save and review games</p>
<p>5. Server side validation of chess moves</p>
<p>6. Security (HTTPS, WSS, CAPTCHA, input validation, password encryption) (partially shown on GitHub)</p>
<p>7. Cron jobs</p>
<p>8. Logging(partially shown on GitHub)</p>
<p>9. Graceful shutdown and recovery of active games</p>
<p>10. Automated deployment of web server to production environment (not shown on GitHub)</p>
<p>11. Highscore board containing top bullet, blitz standard, and correspondence ratings.</p>
<p>12. Configurable user settings, chess piece theme, sound toggle, chat toggle, etc.</p>
<p>13. Database of over 1.8 million chess games, most of which are master/grandmaster games.</p>
<p>14. Player profiles to view rating and past games.</p>
<p>15. Player's name has the country flag where their IP originates from when they register</p>
<p>16. Players can view on going chess games in real time as spectators</p>
<p>17. Lobby and game room chat.</p>
<p>18. Rating history timeline on player's profile using Google Charts.</p>
<p>19. Players can play against cinnamon chess engine and stockfish chess engine (web workers).</p>
<p>20. Web server runs in Windows and Linux.</p>
<p>21. Forum</p>
<p>22. Daily news feed, contains mostly tech articles but it can be configured to get other types of news.</p>
<br>
<p><b>To Do List:</b></p>
<p>1. Chess TV where live games are randomly broadcasted.</p>
<p>2. Admin control</p>
<p>3. Friends list</p>
<p>4. Inbox system</p>
<p>5. Update help, documentation and screenshots</p>
<p>6. Post game analysis with Stockfish engine</p>
<p>7. Statistics page</p>
<p>8. Store zipped source files and databas backup in Google Drive or Dropbox.</p>
<p>9. Read write meta data of each tile to file when saving and loading memory map.</p>
<p>10. Setup algorithm to pick starting player location and save player's map vision to database.</p>
<br>

<b>Special thanks to those who provided amazing third party libaries</b>:
<p>1. chess.js</p>
<p>2. chessboard.js</p>
<p>3. goglicko by Josh Hoak aka Kashomon</p>
<p>4. Tock.js</p>
<p>5. Cinnamon JS Engine</p>
<p>6. download.js v4.1, by dandavis</p>
<p>7. dchest/captcha</p>
<p>8. robfig/cron</p>
<p>9. gopkg.in/gomail.v2 </p>
<p>10. golang.org/x/crypto/scrypt </p>
<p>And many more...</p>

<br>
<b>Other Dependencies</b>:
<p>1. jsonlint</p>
<br>

<b>MUD</b>

Classes:

1. Warrior
2. Barbarian 
3. Monk
4. Mage
5. Thief
6. Ranger
7. Swordmaster
8. Illusionist
9. Priest
10. Necromancer
11. Witch
12. Paladin
13. Alchemist
14. Jester

Breakdown of classes and description:

1. <b>Warriors</b>:
Warriors are well rounded fighter that can use a variety of powerful armor and weaponry. 
Warriors can learn some basic magic.

2. <b>Barbarian</b>: 
Barbarians are ancient warrior who uses brute strength and raw fury
to excel in combat. After many generations of fierce training they have become resistant to magic.

3. <b>Monk</b>:
Monks are masters at martial arts and they can use their spirtual powers to heal and purge disease.

4. <b>Mage</b>:
Mages are spell casters who wield powerful spells. They can do great damage but are physically weak
as a trade off.

5. <b>Thief</b>:
Makes a living by robbing others.

6. <b>Ranger</b>:
Cunning forestmen who lay traps.

7.  <b>Swordmaster</b>:
Swordmasters train with the blade day and night until they become experts.

8.  <b>Illusionist</b>:
Manipulates space and time

9.  <b>Priest</b>:
Holy magic casters

10. <b>Necromancer</b>:
Can raise people from the dead

11. <b>Witch</b>:
Uses dark voodo magic and can create healing potions

12. <b>Paladin</b>:
Holy warriors who can use holy magic
13. <b>Alchemist</b>:
Able to manipulate bodily fluid's and create rare elements
14. <b>Jester</b>:
The utlimate prankster

Races
1. Human
2. Dwarf
3. Elf
4. Troll
5. Ogre
6. Hobbit
7. Vampire

release_notes.txt
2018-04-14
Back end clock not syncing
Allow players to view other live games in real time
Allow unrated games
Add browser cache expire in header
Rematch is missing rated info
Export database to a backup directory
Import database gochess from SQL file if no gochess database exists
Display to player on profile he has no games played if zero games are displayed
Resume game does not display game type of blitz, standard, bullet in profile when game is finished
Pressing the enter key triggers the login button
Integration of an automation build tool
Allow player to choose what piece to promote pawn too
Allow grandmaster games to be viewed from the database
Integrate a daemon to initiate the application instead of using screen
Make clock sync with millisecond precision
Zip or archive database export file to save disk space
Allow database imports and exports be compatible on Linux  machines
User analytics for owner
Fix JS review game bug
Upgrade Golang 1.6 to 1.7 in development and production environment
Piece promotion bug
Pass in localhost or goplaychess.com through command line parsing
Replace error messages with log.Lshortfile
PGN chess notation is not displaying when reviewing old games
Pawn promotion does not work when reviewing old games
Previous and next buttons don't work during a game
Clocks do not sync for spectators
Fix Travis build
Have pgn and status update after each move when reviewing game
Allow players to change board color instead of using default
Upgrade AWS Storage size to be able to store 2 million chess games
Allow random grandmaster game to be searched based on ECO 
Renew goplaychess.com domain name
Automate compression of JS files and FTP to production environment with a single script
Automatically resize images if necessary on startup if they don't meet set requirements
Update images in readme and screenshots
Add country flags to user name when playing a game and in profile
Setup selenium to test load balancing
Search all seeks when a player sends out a seek to see if there is a match
Resolve lobby which disconnects user after a server restart and can be fixed by clearing cache
Allow git pull/sync without keeping history and override local changes in production
Broken image for 404.html at /chess/**
Games restoring
Move unit tests to _unittest folder
Chess Engine
Add forum
Make database, profile, lobby and engine mobile friendly
Renew security certificate
Reset timer for correspondence games
Fix flag displaying as one's own flag instead of other flags on profile
Generate random wilderness map
Automate letsencrypt cert renewal
Make forum usernames clickable and link to profile page if logged in, also allow hover over AJAX
Update github notes to not included closed tickets that were marked with "Wont do"


new.go
// processRegister.go
package gostuff

import (
	"crypto/rand"
	"database/sql"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"html/template"
	"log"
	"net/http"
	"os"
	"time"

	"github.com/dchest/captcha"
	_ "github.com/go-sql-driver/mysql"
	"golang.org/x/crypto/scrypt"
)

// global SessionManager["username"] = sessionID
var SessionManager = make(map[string]string)

//stores rating inforation about user in memory
type Person struct {
	User     string
	Bullet   int16
	Blitz    int16
	Standard int16
}

//process user input when signing in
func ProcessLogin(w http.ResponseWriter, r *http.Request) {
	if r.Method != "POST" {
		w.WriteHeader(404)
		http.ServeFile(w, r, "404.html")
		return
	}
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	userName := template.HTMLEscapeString(r.FormValue("user"))
	passWord := template.HTMLEscapeString(r.FormValue("password"))

	problems, _ := os.OpenFile("logs/errors.txt", os.O_APPEND|os.O_WRONLY, 0666)
	defer problems.Close()
	log := log.New(problems, "", log.LstdFlags|log.Lshortfile)

	capID := template.HTMLEscapeString(r.FormValue("captchaId"))
	capSol := template.HTMLEscapeString(r.FormValue("captchaSolution"))

	if capSol == "" { //then assume user was not displayed captcha

		//check if database connection is open
		if db.Ping() != nil {
			w.Write([]byte("<img src='img/ajax/not-available.png' /> We are having trouble with our server. Please come back later. Error 24"))
			log.Println("DATABASE DOWN!")
			return
		}

		//hashing password
		dk, err1 := scrypt.Key([]byte(passWord), []byte(userName), 16384, 8, 1, 32)
		key := hex.EncodeToString(dk)
		if err1 != nil {
			w.Write([]byte("<img src='img/ajax/not-available.png' /> Something is wrong with the server. Tell admin error 25"))
			log.Println(err1)
			return
		}

		var pass string
		var verify string
		var captcha int
		var email string

		//getting password, verify, captcha and email
		err2 := db.QueryRow("SELECT password, verify, captcha, email FROM userinfo WHERE username=?", userName).Scan(&pass, &verify, &captcha, &email)
		if err2 != nil {
			w.Write([]byte("<img src='img/ajax/not-available.png' /> Wrong username/password combination."))
			log.Println("Incorrect login for ", userName)
			return
		}

		//if user entered password incorrect two times or more then they need to enter captcha to login
		if captcha >= 2 {
			w.Write([]byte("<script>$('#hiddenCap').show();</script><img src='img/ajax/not-available.png' /> You entered password incorrecty too many times. Now you need to enter captcha."))
			return
		}
		//checking if password entered by user matches encrypted key
		if pass != key {
			log.Printf("FAILED LOGIN IP: %s  Method: %s Location: %s Agent: %s\n", r.RemoteAddr, r.Method, r.URL.Path, r.UserAgent())

			if captcha == 1 {
				w.Write([]byte("<script>$('#hiddenCap').show();</script><img src='img/ajax/not-available.png' />  You entered password incorrectly too many times. Now you need to enter captcha."))
			} else {
				w.Write([]byte("<img src='img/ajax/not-available.png' /> Wrong username/password combination."))
			}

			//add 1 to captcha if password was incorrect
			stmt, err := db.Prepare("update userinfo set captcha=? where username=?")
			if err != nil {
				log.Println(err)
				return
			}
			captcha = captcha + 1

			_, err = stmt.Exec(captcha, userName)
			if err != nil {
				log.Println(err)
			}
			return
		}
		if verify != "YES" {
			var tokenInDB string
			var email string

			log.Printf("%s needs to activate his account before logging in.\n", userName)
			w.Write([]byte("<img src='img/ajax/not-available.png' /> You must activate your account by entering the activation token in your email at the activation page. An email has been sent again containing your activation code."))

			//checking if token matches the one entered by user
			err2 := db.QueryRow("SELECT token, email FROM activate WHERE username=?", userName).Scan(&tokenInDB, &email)
			if err2 != nil {
				log.Println(err2)
			} else {
				go func(email, tokenInDB, userName string) {
					Sendmail(email, tokenInDB, userName)
				}(email, tokenInDB, userName)
			}
			return
		}
		// update captcha to zero since login was a sucess
		stmt, err := db.Prepare("update userinfo set captcha=? where username=?")
		if err != nil {
			log.Println(err)
			w.Write([]byte("<img src='img/ajax/not-available.png' /> Error in captcha section 3"))
			return
		}

		_, err = stmt.Exec(0, userName)
		if err != nil {
			log.Println(err)
			w.Write([]byte("<img src='img/ajax/not-available.png' /> Error in captcha section 4"))
			return
		}

		expiration := time.Now().Add(3 * 24 * time.Hour)
		cookie := http.Cookie{Name: "username", Value: userName, Secure: true, HttpOnly: true, Expires: expiration}
		http.SetCookie(w, &cookie)

		//generating random session ID to be stored in the backend
		sessionID := RandomString()

		cookie = http.Cookie{Name: "sessionID", Value: sessionID, Secure: true, HttpOnly: true, Expires: expiration}
		http.SetCookie(w, &cookie)

		SessionManager[userName] = sessionID

		w.Write([]byte("<script>window.location = '/memberHome'</script>"))

	} else if !captcha.VerifyString(capID, capSol) {
		w.Write([]byte("<script>document.getElementById('captchaSolution').value = '';</script><img src='img/ajax/not-available.png' /> Wrong captcha solution! Please try again."))
	} else {
		//check if database connection is open
		if db.Ping() != nil {
			w.Write([]byte("<img src='img/ajax/not-available.png' /> We are having trouble with our server. Please come back later. Error 24"))
			log.Println("DATABASE DOWN!")
			return
		}

		//hashing password
		dk, err1 := scrypt.Key([]byte(passWord), []byte(userName), 16384, 8, 1, 32)
		key := hex.EncodeToString(dk)
		if err1 != nil {
			w.Write([]byte("<img src='img/ajax/not-available.png' /> Something is wrong with the server. Tell admin error 25"))
			log.Println(err1)
			return
		}

		var pass string
		var verify string
		var captcha int
		var email string
		var token string

		//getting password, verify, captcha, email from database
		err2 := db.QueryRow("SELECT password, verify, captcha, email FROM userinfo WHERE username=?", userName).Scan(&pass, &verify, &captcha, &email)
		if err2 != nil {
			w.Write([]byte("<img src='img/ajax/not-available.png' /> Wrong username/password combination."))
			//check if there was more then one incorrect login attempt
			log.Println(err2)
			return
		}

		if captcha == 5 {
			w.Write([]byte("<img src='img/ajax/not-available.png' /> This account has been deactivated becaue too many incorrect login attempts were made. An email has been sent to you on instructions on reactivating your account."))

			//increment captcha counter
			stmt, err := db.Prepare("update userinfo set captcha=? where username=?")
			if err != nil {
				log.Println(err)
				return
			}
			captcha = captcha + 1

			_, err = stmt.Exec(captcha, userName)
			if err != nil {
				log.Println(err)
				return
			}

			//create activation token in database and send user notifying them that their was five incorrect login attempts
			stmt, err = db.Prepare("INSERT activate SET username=?, token=?, email=?, expire=?")
			if err != nil {
				log.Println(err)
				return
			}

			token = RandomString()
			_, err = stmt.Exec(userName, token, email, time.Now())
			if err != nil {
				log.Println(err)
				return
			}
			//sends email to user with the token activation
			go func(email, token, userName, address string) {
				SendAttempt(email, token, userName, address)
			}(email, token, userName, r.RemoteAddr)
			return

		} else if captcha > 5 { //tell user on the front end that this account has too many login attempts, resends activation token
			w.Write([]byte("<img src='img/ajax/not-available.png' /> This account has been deactivated because too many incorrect login attempts were made. An email has been sent again regarding how to reactivate your account."))

			err2 := db.QueryRow("SELECT token FROM activate WHERE username=?", userName).Scan(&token)
			if err2 != nil {
				log.Println(err2)
				return
			}

			//sends email again to user with the token activation
			go func(email, token, userName, address string) {
				SendAttempt(email, token, userName, address)
			}(email, token, userName, r.RemoteAddr)

			return
		}
		if pass != key {
			w.Write([]byte("<img src='img/ajax/not-available.png' /> Wrong username/password combination."))
			log.Printf("FAILED LOGIN IP: %s  Method: %s Location: %s Agent: %s\n", r.RemoteAddr, r.Method, r.URL.Path, r.UserAgent())

			//add 1 to captcha if password was incorrect
			stmt, err := db.Prepare("update userinfo set captcha=? where username=?")
			if err != nil {
				log.Println(err)
			}
			captcha = captcha + 1

			_, err = stmt.Exec(captcha, userName)
			if err != nil {
				log.Println(err)
			}
			return
		}
		if verify != "YES" {
			var tokenInDB string
			var email string

			log.Printf("%s needs to activate his account before logging in.\n", userName)
			w.Write([]byte("<img src='img/ajax/not-available.png' /> You must activate your account by entering the activation token in your email at the activation page. An email has been sent again containing your activation code."))

			//checking if token matches the one entered by user
			err2 := db.QueryRow("SELECT token, email FROM activate WHERE username=?", userName).Scan(&tokenInDB, &email)
			if err2 != nil {
				log.Println(err2)
			} else {
				go func(email, tokenInDB, userName string) {
					Sendmail(email, tokenInDB, userName)
				}(email, tokenInDB, userName)
			}
			return
		}

		// update captcha to zero since login was a sucess
		stmt, err := db.Prepare("update userinfo set captcha=? where username=?")
		if err != nil {
			log.Println(err)
		}

		_, err = stmt.Exec(0, userName)
		if err != nil {
			log.Println(err)
		}

		expiration := time.Now().Add(3 * 24 * time.Hour)
		cookie := http.Cookie{Name: "username", Value: userName, Secure: true, HttpOnly: true, Expires: expiration}
		http.SetCookie(w, &cookie)

		//generating random session ID to be stored in the backend
		sessionID := RandomString()

		cookie = http.Cookie{Name: "sessionID", Value: sessionID, Secure: true, HttpOnly: true, Expires: expiration}
		http.SetCookie(w, &cookie)

		SessionManager[userName] = sessionID

		w.Write([]byte("<script>window.location = '/memberHome'</script>"))
	}
}

//processes the users input when signing up
func ProcessRegister(w http.ResponseWriter, r *http.Request) {

	if r.Method != "POST" {
		w.WriteHeader(404)
		http.ServeFile(w, r, "404.html")
		return
	}
	w.Header().Set("Content-Type", "text/html; charset=utf-8")

	if !captcha.VerifyString(template.HTMLEscapeString(r.FormValue("captchaId")), template.HTMLEscapeString(r.FormValue("captchaSolution"))) {
		w.Write([]byte("<script>document.getElementById('captchaSolution').value = '';</script><img src='img/ajax/not-available.png' /> Wrong captcha solution"))

	} else {

		userName := template.HTMLEscapeString(r.FormValue("username"))
		passWord := template.HTMLEscapeString(r.FormValue("pass"))
		confirm := template.HTMLEscapeString(r.FormValue("confirm"))
		email := template.HTMLEscapeString(r.FormValue("email"))

		if len(userName) < 3 || len(userName) > 12 {
			w.Write([]byte("<img src='img/ajax/not-available.png' /> Please choose a username between 3 and 12 characters long."))

		} else if passWord != confirm {
			w.Write([]byte("<img src='img/ajax/not-available.png' /> Your password and confirm password did not match"))

		} else if len(passWord) < 5 || len(passWord) > 32 {
			w.Write([]byte("<img src='img/ajax/not-available.png' /> Password must be at between 5 to 32 characters long"))

		} else if len(email) < 5 || len(email) > 30 {

			w.Write([]byte("<img src='img/ajax/not-available.png' /> Please choose an email between 5 and 30 characters long"))

		} else {

			problems, err := os.OpenFile("logs/errors.txt", os.O_APPEND|os.O_WRONLY, 0666)
			defer problems.Close()
			log := log.New(problems, "", log.LstdFlags|log.Lshortfile)

			//check if database connection is open
			if db.Ping() != nil {
				w.Write([]byte("<img src='img/ajax/not-available.png' /> We are having trouble with our server. Please come back later. Report to admin Error 27"))
				log.Println("DATABASE DOWN!")
				return
			}

			//check if username already exists, if it does then break out and inform user
			//use javascript for as well as check this in backend in Golang
			var name string
			//checking if name exists
			checkName := db.QueryRow("SELECT username FROM userinfo WHERE username=?", userName).Scan(&name)

			if checkName != sql.ErrNoRows {
				w.Write([]byte("<img src='img/ajax/not-available.png' /> Username already exist. Please choose another username"))
				log.Printf("Prevented host %s from choosing duplicate username %s\n", r.RemoteAddr, userName)
				return

			}
			message := "<script>$('#register').hide();</script><img src='img/ajax/available.png' /> Hello " + userName + "! Please check email for instructions to verify your account."
			//if reached here just notify user to check his email and continue on with the account creation
			w.Write([]byte(message))

			//hashing password
			dk, err1 := scrypt.Key([]byte(passWord), []byte(userName), 16384, 8, 1, 32)
			key := hex.EncodeToString(dk)
			if err1 != nil {
				w.Write([]byte("<img src='img/ajax/not-available.png' /> We are having trouble with our server. Please come back later. Report to admin Error 28"))
				log.Println(err)
				return
			}

			//inserting into database
			stmt, err := db.Prepare("INSERT userinfo SET username=?, password=?, email=?, date=?, time=?, host=?, verify=?, captcha=?")
			if err != nil {
				w.Write([]byte("<img src='img/ajax/not-available.png' /> We are having trouble with our server. Please come back later. Report to admin Error 29"))
				log.Println(err)
				return
			}

			date := time.Now()

			res, err := stmt.Exec(userName, key, email, date, date, r.RemoteAddr, "NO", 0)
			if err != nil {
				w.Write([]byte("<img src='img/ajax/not-available.png' /> We are having trouble with our server. Please come back later. Report to admin Error 30"))
				log.Println(err)
				return
			}

			id, err := res.LastInsertId()
			if err != nil {
				w.Write([]byte("<img src='img/ajax/not-available.png' /> We are having trouble with our server. Please come back later. Report to admin Error 31"))
				log.Println(err)
				return
			}
			log.Printf("Account %s and id %d was created in userinfo table.\n", userName, id)

			token := RandomString()

			//preparing token activation
			stmt, err = db.Prepare("INSERT activate SET username=?, token=?, email=?, expire=?")
			if err != nil {
				w.Write([]byte("<img src='img/ajax/not-available.png' /> We are having trouble with our server. Please come back later. Report to admin Error 32"))
				log.Println(err)
				return
			}

			res, err = stmt.Exec(userName, token, email, date)
			if err != nil {
				w.Write([]byte("<img src='img/ajax/not-available.png' /> We are having trouble with our server. Please come back later. Report to admin Error 33"))
				log.Println(err)
				return
			}
			//sends email to user
			go func(email, token, userName string) {
				Sendmail(email, token, userName)
			}(email, token, userName)

			//setting up player's rating
			stmt, err = db.Prepare("INSERT rating SET username=?, bullet=?, blitz=?, standard=?, bulletRD=?, blitzRD=?, standardRD=?")
			if err != nil {
				w.Write([]byte("<img src='img/ajax/not-available.png' /> We are having trouble with our server. Please come back later. Report to admin Error 34"))
				log.Println(err)
				return
			}

			res, err = stmt.Exec(userName, "1500", "1500", "1500", "350.0", "350.0", "350.0")
			if err != nil {
				w.Write([]byte("<img src='img/ajax/not-available.png' /> We are having trouble with our server. Please come back later. Report to admin Error 35"))
				log.Println(err)
				return
			}
		}
	}
}

func CheckUserName(w http.ResponseWriter, r *http.Request) {
	if r.Method == "POST" {
		userName := template.HTMLEscapeString(r.FormValue("username"))

		//making sure username fits length requirement
		if len(userName) < 3 || len(userName) > 12 {
			return
		}

		//check if database connection is open
		if db.Ping() != nil {
			fmt.Printf("ERROR 2 PINGING DB IP: %s \n", r.RemoteAddr)
			w.Write([]byte("<img src='img/ajax/not-available.png' /> Please come back later."))
			return
		}

		var name string
		//checking if name exists
		checkName := db.QueryRow("SELECT username FROM userinfo WHERE username=?", userName).Scan(&name)
		switch {
		case checkName == sql.ErrNoRows:
			w.Write([]byte(" <img src='img/ajax/available.png' /> Username available"))
			fmt.Printf("Username %s is available.\n", userName)
		case checkName != nil:
			fmt.Printf("ERROR 3 CHECKNAME IP is %s\n", r.RemoteAddr)
		default:
			w.Write([]byte("<img src='img/ajax/not-available.png' /> Username taken"))
		}
	}
}

func RandomString() string {
	size := 12 // change the length of the generated random string here 12 is actually 16

	rb := make([]byte, size)
	_, err := rand.Read(rb)

	if err != nil {
		fmt.Println("new.go RandomString 1 ", err)
	}

	token := base64.URLEncoding.EncodeToString(rb)
	return token
}
